### PR TITLE
Support VPA

### LINKED ISSUE #462: Support VPA

### PR DESCRIPTION
**Pull request by jiangpengcheng**
Fixes #462

### Motivation

*Explain here the context, and why you're making that change. What is the problem you're trying to solve.*

### Modifications

Support VPA(vertical pod autoscaler)

### Verifying this change

- [x] Make sure that the change passes the CI checks.

- [x] This change added tests and can be verified as follows:

  - *Added integration tests for function with VPA enabled

### Documentation

Check the box below.

Need to update docs? 

- [x] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [ ] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

File: controllers/function.go [hunk 1]
```diff
@@ -228,6 +229,34 @@ func (r *FunctionReconciler) ApplyFunctionHPA(ctx context.Context, function *v1a
...
+		APIVersion: function.APIVersion,
+	}
+
+	err := applyVPA(ctx, r, r.Log, condition, objectMeta, targetRef, function.Spec.Pod.VPA, "function", function.Namespace, function.Name)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (jiangpengcheng) [reply controllers/function.go:258]: do you mean to check whether the `VPASpec` is nil or not?

[the remark this answers is not in the document]
Comment (tpiperatgod) [reply controllers/function.go:258]: yes

[the remark this answers is not in the document]
Comment (jiangpengcheng) [reply controllers/function.go:258]: but if `VPASpec` is nil, we may still need to delete the old VPA

[the remark this answers is not in the document]
Comment (tpiperatgod) [reply controllers/function.go:258]: can this be done by conditional checking? I think HPA may also need this operation...

[the remark this answers is not in the document]
Comment (jiangpengcheng) [reply controllers/function.go]: I think it's better to use an another pr to add same logic for HPA, I will do it later

**Review by tpiperatgod — DISMISSED** [review]
lgtm

**Comment by nlu90** [comment]
Let's have a discussion during SIG

**Review by freeznet — APPROVED** [review]

File: charts/function-mesh-operator/Chart.yaml [hunk 2]
```diff
@@ -34,3 +34,6 @@ dependencies:
     condition: admissionWebhook.enabled
     version: 0.2.6
 #    repository: https://charts.functionmesh.io/
+  - name: vertical-pod-autoscaler
+    repository: https://cowboysysop.github.io/charts/
```
Comment (freeznet) [inline charts/function-mesh-operator/Chart.yaml:38]: Is there any official charts for VPA? I noticed that the one this PR uses is not an official one, see https://github.com/cowboysysop/charts/tree/master/charts/vertical-pod-autoscaler

**Review by freeznet — APPROVED** [review]

File: charts/function-mesh-operator/Chart.yaml (hunk 2 above)
Comment (jiangpengcheng) [reply charts/function-mesh-operator/Chart.yaml:38]: there is no official charts, the official are using shell scripts to install vpa

**Pull request closed without merging**